dali/kernels/shape.h

```diff
@@ -0,0 +1,299 @@
+      : shapes(std::accumulate(sample_shapes.begin(), sample_shapes.end(), std::vector<int64_t>(),
+                               [](std::vector<int64_t> current, const TensorShape<sample_dim_> &next) {
+                                 return std::move(current).insert(current.end(), next.shape.begin(),
+                                                                  next.shape.end());
+                               })) {}
+
+  template <int tensor_dim = sample_dim_>
+  typename std::enable_if<tensor_dim == sample_dim_ || tensor_dim == DynamicTensorShape,
+                          TensorShape<tensor_dim>>::type
+  tensor_shape(int64_t sample) const {
+    TensorShape<tensor_dim> out;
+    if (tensor_dim == DynamicTensorShape) {
+      out.shape.resize(sample_dim());
+    }
+    int64_t base = sample_dim() * sample;
+    for (int i = 0; i < sample_dim(); i++) {
+      out[i] = shapes[base + i];
+    }
+    return out;
+  }
+
+  span<int64_t, sample_dim_> tensor_shape_span(int64_t sample) const {
+    return {&shapes[sample * sample_dim()], sample_dim()};
+  }
+
+  constexpr int sample_dim() const { return sample_dim_; }
+
+  int samples() const { return shapes.size() / sample_dim(); }
+
+  std::vector<int64_t> shapes;
+};
+
+template <>
+struct TensorListShape<DynamicTensorShape> {
+  template <int tensor_dim>
+  TensorShape<tensor_dim> tensor_shape(int64_t sample) const {
+    TensorShape<tensor_dim> out;
+    if (tensor_dim == DynamicTensorShape) {
+      out.shape.resize(sample_dim());
+    }
+    int64_t base = sample_dim() * sample;
+    for (int i = 0; i < sample_dim(); i++) {
+      out[i] = shapes[base + i];
+    }
+    return out;
+  }
+
+  span<int64_t> tensor_shape_span(int64_t sample) const {
+    return {&shapes[sample * sample_dim()], static_cast<size_t>(sample_dim())};
+  }
+
+  int sample_dim() const { return dim; }
+
+  int samples() const { return shapes.size() / sample_dim(); }
+
+  std::vector<int64_t> shapes;
+  int dim;
+};
+
+template <int sample_dim>
+std::vector<ptrdiff_t> calculate_offsets(const TensorListShape<sample_dim> &tls) {
+  std::vector<ptrdiff_t> offsets;
+  offsets.resize(tls.samples() + 1);
+  offsets.push_back(0);
+  for (int i = 0; i < tls.samples(); i++) {
+    auto sample_shape_span = tls.tensor_shape_span(i);
+    auto off = std::accumulate(sample_shape_span.begin(), sample_shape_span.end(),
+                               static_cast<ptrdiff_t>(1),
+                               [](const ptrdiff_t &a, const ptrdiff_t &b) { return a * b; });
+    offsets.push_back(off);
+  }
+  return offsets;
+}
+
+// TODO:
+// * `multiply(span)`
+// * `flatten` to vector
+// * `T* at(...)` as free function
+
+
+template <int sample_dim_>
+struct TensorListDim {
+  inline constexpr int sample_dim() const { return sample_dim_; }
+
+ protected:
+  inline void set_sample_dim(int dim) { assert(dim == sample_dim_); }
+};
+
+template <>
+struct TensorListDim<-1> {
```

Comment:
What does -1 represent here? If it carries special meaning (e.g. "dynamic" or something) I think it'd be beneficial to make a named constant for it